### PR TITLE
blog: Add GSoC'24 post on testing mimalloc memory allocator

### PR DESCRIPTION
**Pull request by huyang531**
Signed-off-by: Yang Hu [yanghuu531@gmail.com](mailto:yanghuu531@gmail.com)

### CODE REVIEW

File: content/blog/2024-07-12-unikraft-gsoc-test-mimalloc-on-unikraft.mdx [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+After that, I will test mimalloc with other more complicated benchmarks, which may include:
+
+- `threadtest`: tests general scalability of the allocator;
+Each thread repeatedly allocates a number of number of objects, does a fixed amount of computation, and then deletes its objects.
+- `linux-scalability`: a basic scalability test from Linux libc malloc testing, similar to `threadtest` except that there is no extra 'work' between allocations and frees so contention may be higher.
+- `cache-thrash`: tests for active false sharing
+- `larson`: simulates a server by having each thread allocate and deallocate objects and then transfer some objects to other threads to be freed
+- `phong` - a benchmark that randomly selects allocation sizes, and randomly chooses an allocated item to free
+
+Once fully tested, I will continue on with the work items outlined in [my previous post].(https://unikraft.org/blog/2024-06-16-unikraft-gsoc-port-mimalloc-to-unikraft#next-steps)
```
Comment (StefanJum) [inline content/blog/2024-07-12-unikraft-gsoc-test-mimalloc-on-unikraft.mdx:134]:
```suggestion
Once fully tested, I will continue on with the work items outlined in [my previous post](https://unikraft.org/blog/2024-06-16-unikraft-gsoc-port-mimalloc-to-unikraft#next-steps).
```

**Review by StefanJum — CHANGES_REQUESTED** [review]
See the inline comment.
Besides that, please sign-off your commit.

**Comment by StefanJum** [comment]
@RaduNichita @razvanvirtan can you also take a look from a technical point of view?

**Comment by StefanJum** [comment]
@huyang531 squash the commits, don't add the co-authored-by tag.

File: content/blog/2024-07-12-unikraft-gsoc-test-mimalloc-on-unikraft.mdx [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+The steps of porting the memory allocators are as follows:
+
+1. Adapt the existing port, which uses `lib-newlib` and `lib-pthread-embedded` as the C library to use `lib-musl` instead
+1. Patch the existing port of mimalloc (v1.6.3) to adapt to the latest Unikraft's memory allocation interface
+1. Patch the latest mimalloc (v2.1.7) to the latest Unikraft version (v0.17.0)
+
+## Current Progress
+
+Last month, I successfully ported the mimalloc memory allocator to Unikraft, marked by a successful compilation of mimalloc v1.6
+.3
```
Comment (RaduNichita) [inline content/blog/2024-07-12-unikraft-gsoc-test-mimalloc-on-unikraft.mdx:43]: Put on previous line

File: content/blog/2024-07-12-unikraft-gsoc-test-mimalloc-on-unikraft.mdx [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+against the latest Unikraft core (v2.1.7) with `lib-musl`.
```
Comment (RaduNichita) [inline content/blog/2024-07-12-unikraft-gsoc-test-mimalloc-on-unikraft.mdx:44]:
```suggestion
against the latest Unikraft core (v2.1.7) with `lib-musl`.
```

Should it be v0.17.0?

File: content/blog/2024-07-12-unikraft-gsoc-test-mimalloc-on-unikraft.mdx [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+Following that, this month I am focused on stress-testing the memory allocator to validate its usability and correctness.
```
Comment (RaduNichita) [inline content/blog/2024-07-12-unikraft-gsoc-test-mimalloc-on-unikraft.mdx:45]:
```suggestion
Following that, this month I am focused on stress-testing the `mimalloc` memory allocator to validate its usability and correctness.
```

File: content/blog/2024-07-12-unikraft-gsoc-test-mimalloc-on-unikraft.mdx [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+
+### Benchmark
+
+Right after porting, I tested the allocator by running a few `malloc()` and `free()` calls in a single threaded `main()` function.
```
Comment (RaduNichita) [inline content/blog/2024-07-12-unikraft-gsoc-test-mimalloc-on-unikraft.mdx:49]:
```suggestion
After porting was completed, I tested the allocator by running a few `malloc()` and `free()` calls in a single threaded `main()` function.
```

File: content/blog/2024-07-12-unikraft-gsoc-test-mimalloc-on-unikraft.mdx [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+It creates a number of worker threads that allocate a given-sized object, repeatedly write on it, then free it.
+An allocator that allows multiple threads to re-use the same small object (possibly all in one cache-line) will scale poorly, while an allocator like mimalloc, which gives each thread a private and page-aligned heap, will exhibit near-linear performance scaling.
+
+### Porting `cache-scratch`
+
+It is fairly intuitive to port the C program itself to Unikraft as the Musl library already provides most of what we need, including the `pthread` interface.
+It would be ideal if we could separate the helper functions from the main logic and build the benchmark as a library (as we do need to test the memory allocator against more benchmarks in the future), because then we could just build an all-in-one image and run a large number of tests with different benchmarks and different parameters using an automated script.
+However, I haven't figured out a good way to do it as [Unikraft's build system](https://unikraft.org/docs/internals/build-system#makefileuk) is quite complex.
+So for now, the benchmark is ported as a single `main.c` file that will run as an application.
```
Comment (RaduNichita) [inline content/blog/2024-07-12-unikraft-gsoc-test-mimalloc-on-unikraft.mdx:60]: Do you have some preliminary results to put into the blogpost?

File: content/blog/2024-07-12-unikraft-gsoc-test-mimalloc-on-unikraft.mdx [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+- `phong` - a benchmark that randomly selects allocation sizes, and randomly chooses an allocated item to free
+
+Once fully tested, I will continue on with the work items outlined in [my previous post](https://unikraft.org/blog/2024-06-16-unikraft-gsoc-port-mimalloc-to-unikraft#next-steps).
+
+Meanwhlie, I will also try to debug the `test_cmp_requeue_two_waiters_no_requeue` to the best of my ability.
+
+## Acknowledgements
+
+Special thanks to Răzvan Vîrtan and Radu Nichita, my two amazing mentors, for their support along the way.
+I would also like to thank Razvan Deaconescu, Hugo Lefeuvre, Ștefan Jumarea, Marco Schlumpp, Sergiu Moga, and the entire Unikraft community for all the work, discussions, and guidance which made this project possible.
```
Comment (RaduNichita) [inline content/blog/2024-07-12-unikraft-gsoc-test-mimalloc-on-unikraft.mdx:141]:
```suggestion
I would also like to thank Răzvan Deaconescu, Hugo Lefeuvre, Ștefan Jumarea, Marco Schlumpp, Sergiu Moga, and the entire Unikraft community for all the work, discussions, and guidance which made this project possible.
```

File: content/blog/2024-07-12-unikraft-gsoc-test-mimalloc-on-unikraft.mdx [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+
+Last month, I successfully ported the mimalloc memory allocator to Unikraft, marked by a successful compilation of mimalloc v1.6
+.3
+against the latest Unikraft core (v2.1.7) with `lib-musl`.
+Following that, this month I am focused on stress-testing the memory allocator to validate its usability and correctness.
+
+### Benchmark
+
+Right after porting, I tested the allocator by running a few `malloc()` and `free()` calls in a single threaded `main()` function.
+That having no problem, I ported `cache-scratch`, a multi-threaded memory allocation benchmark, to Unikraft.
```
Comment (RaduNichita) [inline content/blog/2024-07-12-unikraft-gsoc-test-mimalloc-on-unikraft.mdx:50]: Make a new paragraph with this

**Review by RaduNichita — COMMENTED** [review]
Please see my comments @huyang531

File: content/blog/2024-07-12-unikraft-gsoc-test-mimalloc-on-unikraft.mdx (hunk 6 above)
Comment (huyang531) [reply content/blog/2024-07-12-unikraft-gsoc-test-mimalloc-on-unikraft.mdx:60]: I do have the code ready - where should I put them?

File: content/blog/2024-07-12-unikraft-gsoc-test-mimalloc-on-unikraft.mdx (hunk 3 above)
Comment (huyang531) [reply content/blog/2024-07-12-unikraft-gsoc-test-mimalloc-on-unikraft.mdx:44]: Yes it should 👍

File: content/blog/2024-07-12-unikraft-gsoc-test-mimalloc-on-unikraft.mdx (hunk 2 above)
Comment (huyang531) [reply content/blog/2024-07-12-unikraft-gsoc-test-mimalloc-on-unikraft.mdx:43]: The Linter has a "one end of sentence punctuation per line" rule so it doesn't want to see two `.`s in one line... Do you know of any way to override that?

Comment (huyang531) [reply content/blog/2024-07-12-unikraft-gsoc-test-mimalloc-on-unikraft.mdx]: Nvm - I put them in a code block and it works just fine!

File: content/blog/2024-07-12-unikraft-gsoc-test-mimalloc-on-unikraft.mdx (hunk 6 above)
Comment (StefanJum) [reply content/blog/2024-07-12-unikraft-gsoc-test-mimalloc-on-unikraft.mdx:59]: Anywhere. You can create a repo on your github, or a gist or anything, lust add a link here.

**Comment by StefanJum** [comment]
@huyang531 use `git config --global user.name "Your Name"` and then amend the commit.
Right now it shows
```
Author: huyang531 <yanghuu531@gmail.com>
Commit: huyang531 <yanghuu531@gmail.com>
```

It should be your name and email.

File: content/blog/2024-07-12-unikraft-gsoc-test-mimalloc-on-unikraft.mdx (hunk 6 above)
Comment (huyang531) [reply content/blog/2024-07-12-unikraft-gsoc-test-mimalloc-on-unikraft.mdx]: Done

**Review by StefanJum — APPROVED** [review]
Reviewed-by: Stefan Jumarea <stefanjumarea02@gmail.com>
Approved-by: Stefan Jumarea <stefanjumarea02@gmail.com>

**Pull request merged**